core/src/raw/oio/read/buffer_reader.rs

```diff
@@ -0,0 +1,457 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use bytes::BufMut;
+use bytes::{Bytes, BytesMut};
+
+use tokio::io::ReadBuf;
+
+use std::cmp::min;
+use std::io::SeekFrom;
+
+use std::task::ready;
+use std::task::Context;
+use std::task::Poll;
+
+use crate::raw::*;
+use crate::*;
+
+/// [BufferReader] allows the underlying reader to fetch data at the buffer's size
+/// and is used to amortize the IO's overhead.
+pub struct BufferReader<R> {
+    r: R,
+    cur: u64,
+
+    buf: BytesMut,
+    cap: usize,
+    /// Indicate buffered data range start.
+    /// `None` stand for no buffered data.
+    buf_start: Option<usize>,
+}
+
+impl<R> BufferReader<R> {
+    /// Create a new [`oio::Reader`] with a buffer.
+    ///
+    /// # Input
+    ///
+    /// The input is an Accessor will may return a seekable reader.
+    pub fn new(r: R, cap: usize) -> BufferReader<R> {
+        BufferReader {
+            r,
+            cur: 0,
+
+            // Lazy allocation.
```

Comment:
This lazy allocation doesn't make sure: the cap of this buf won't be changed during entire lifetime.